saleor/order/actions.py

```diff
@@ -151,23 +152,23 @@ def fulfillment_tracking_updated(
     get_plugins_manager().order_updated(fulfillment.order)
 
 
-def cancel_fulfillment(fulfillment: "Fulfillment", user: "User", restock: bool):
+def cancel_fulfillment(
+    fulfillment: "Fulfillment", user: "User", warehouse: "Warehouse"
+):
     """Cancel fulfillment.
 
-    Return products to corresponding stocks if restock is set to True.
+    Return products to corresponding stocks.
     """
     events.fulfillment_canceled_event(
         order=fulfillment.order, user=user, fulfillment=fulfillment
     )
-    if restock:
-        events.fulfillment_restocked_items_event(
-            order=fulfillment.order, user=user, fulfillment=fulfillment
-        )
-        restock_fulfillment_lines(fulfillment)
-    for line in fulfillment:
-        order_line = line.order_line
-        order_line.quantity_fulfilled -= line.quantity
-        order_line.save(update_fields=["quantity_fulfilled"])
+    events.fulfillment_restocked_items_event(
+        order=fulfillment.order,
+        user=user,
+        fulfillment=fulfillment,
+        warehouse_pk=warehouse.pk,
+    )
+    restock_fulfillment_lines(fulfillment, warehouse)
```

Comment:
Maybe we should move this function up, and the events down -> it would be better if they are created after `restock_fulfillment_lines` is successfully done, otherwise we would create events even if it failed for some reason.